website/static/js/nodesPrivacy.js

```diff
@@ -0,0 +1,281 @@
+
+var MESSAGES = {
+    makeProjectPublicWarning:
+                        'Please review your projects, components, and add-ons for sensitive or restricted information before making them public.' +
+                        '<br><br>Once they are made public, you should assume they will always be public. You can ' +
+                        'return them to private later, but search engines (including Google’s cache) or others may access files before you do.',
+
+    selectNodes: 'Adjust your privacy settings by checking the boxes below. ' +
+                        '<br><br><b>Checked</b> projects and components will be <b>public</b>.  <br><b>Unchecked</b> components will be <b>private</b>.',
+    confirmWarning: {
+        nodesPublic: 'The following projects and components will be made <b>public</b>.',
+        nodesPrivate: 'The following projects and components will be made <b>private</b>.'
+    }
+};
+
+    /**
+     * take treebeard tree structure of nodes and get a dictionary of parent node and all its
+     * children
+     */
+function getNodesOriginal(nodeTree, nodesOriginal) {
+    var i;
+    var nodeId = nodeTree.node.id;
+    nodesOriginal[nodeId] = {
+        public: nodeTree.node.is_public,
+        id: nodeTree.node.id,
+        title: nodeTree.node.title,
+        canWrite: nodeTree.node.can_write,
+        changed: false
+    };
+
+    if (nodeTree.children) {
+        for (i in nodeTree.children) {
+            nodesOriginal = getNodesOriginal(nodeTree.children[i], nodesOriginal);
+        }
+    }
+    return nodesOriginal;
+}
+
+    /**
+     * patches all the nodes in a changed state
+     * uses API v2 bulk requests
+     */
+function patchNodesPrivacy(nodes) {
+    var nodesV2Url = window.contextVars.apiV2Prefix + 'nodes/';
+    var nodesPatch = nodes.splice(0,10).map(function (node) {
+        return {
+            'type': 'nodes',
+            'id': node.id,
+            'attributes': {
+                'public': node.public
+            }
+        };
+    });
+    return $.ajax({
+        url: nodesV2Url,
+        type: 'PATCH',
+        dataType: 'json',
+        contentType: 'application/vnd.api+json; ext=bulk',
+        crossOrigin: true,
+        xhrFields: {withCredentials: true},
+        processData: false,
+        data: JSON.stringify({
+            data: nodesPatch
+        })
+    }).done(function (response) {
+        if (nodes.length === 0) {
+            return;
+        }
+        return patchNodesPrivacy(nodes);
+    }).fail(function(xhr, status, error) {
+        $osf.growl('Error', 'Unable to update project privacy');
+        Raven.captureMessage('Could not PATCH project settings.', {
+            url: nodesV2Url, status: status, error: error
+        });
+    });
+}
+
+    /**
+     * view model which corresponds to nodes_privacy.mako (#nodesPrivacy)
+     *
+     * @type {NodesPrivacyViewModel}
+     */
+var NodesPrivacyViewModel = function(data, parentIsPublic) {
```

Comment:
This class does not need most of what's in `data`. Also, if the format of `data` changes, it could break here; it's probably best to just pass the URL for fetching the node tree.